transcript/src/main/java/com/transcriptanalyzer/transcript/Tree.java

```diff
@@ -71,17 +78,95 @@ public Node hasChild(String intent){
+        public LinkedHashMap<String, Integer> getBestLeafData() {
+            // create a list for the best leaf data of all children
+            LinkedHashMap<String, Integer> leafData = new LinkedHashMap<>();
+            // create a placeholder list for the top three leaf intents
+            LinkedHashMap<String, Integer> bestLeafData = new LinkedHashMap<>();
+
+            // if this is a leaf node (has no children), set the best leaf data to its own intent and occurrences
+            // so that it may be returned
+            if (this.children.isEmpty()) {
+                bestLeafData.put(this.intent, this.occurrences);
+            }
+            // otherwise, get the best 3 intents and their occurrences from this node's children
+            else {
+                // for each child, get its best leaf data and merge it all into leafData
+                for (Node child: this.children) {
+                    // take the best leaf data of the current child
+                    LinkedHashMap<String, Integer> childBestLeaves = child.getBestLeafData();
+
+                    // for each entry of this child's best leaf data, merge the entry into leafData
+                    for (Map.Entry<String, Integer> leaf: childBestLeaves.entrySet()) {
+                        // get the intent and occurrence of the current entry
+                        String intent = leaf.getKey();
+                        Integer occurrences = leaf.getValue();
+
+                        // merge it into leaf data. If an occurrence is new, it is added to the LinkedHashMap.
+                        // If it already exists, their values are summed using Integer::sum
+                        leafData.merge(intent, occurrences, Integer::sum);
+                    }
+                }
+
+                // move the leaves into a list
+                List<Map.Entry<String, Integer>> leafDataEntries = new ArrayList<>(leafData.entrySet());
+
+                // sort the list using the Map.Entry comparing by value comparator in reverse order
+                // this sorts each hash map entry, which is stored in a list, in descending order
+                leafDataEntries.sort(Map.Entry.<String, Integer>comparingByValue().reversed());
+
+                // determine how many entries to take from the list
+                // take the last index of the entries we will take. no more than 3,
+                // and no more than the number of elements in the list
+                int last_index = java.lang.Math.min(2, leafDataEntries.size() - 1);
+
+                // for each index of the entries we want
+                for (int i = 0; i <= last_index; i++) {
+                    // directly extract the key and value at that index
+                    String key = leafDataEntries.get(i).getKey();
+                    Integer value = leafDataEntries.get(i).getValue();
```

Comment:
This is a really clever way of isolating the top intents while avoiding potential errors - well done!